### PR TITLE
refactor pre-selected coin code

### PR DESCRIPTION
**Pull request by instagibbs**
This section of code loops over the same coins twice, in two different ways depending on whether other inputs are allowed, even though the net effect is just an early return in one case. One used the available coins list(pruned to just preselected coins) and the other the preselected coins directly. Instead just do it once, and return early if no other inputs are allowed.

Note that by definition coin control is in use, so "fSpendable" will always be true, so filtering for this is a NOP.

Added basic unit test case as well.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2537,6 +2523,11 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
             return false; // TODO: Allow non-wallet inputs
     }
 
+    if (coin_control.HasSelected() && !coin_control.fAllowOtherInputs)
+    {
+        return (nValueRet >= nTargetValue);
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: What about `setCoinsRet`?

Comment (Empact) [reply src/wallet/wallet.cpp]: `nValueRet` is never updated, so this will always return false.

Comment (laanwj) [reply src/wallet/wallet.cpp]: So it will always fail when `coin_control.fAllowOtherInputs` is not set?

**Comment by Empact** [comment]
NACK 4aaa5ba - this code isn't redundant, the out args are used by caller `CWallet::CreateTransaction`

**Comment by Empact** [comment]
The fact that the test suite passes with this removal is a smell though - how about a test for `CreateTransaction` or `SelectCoins` that fails with this removal?

**Comment by instagibbs** [comment]
I will write tests fix this, thanks. Just not at my computer today.

**Comment by laanwj** [comment]
> The fact that the test suite passes with this removal is a smell though - how about a test for CreateTransaction or SelectCoins that fails with this removal?

This is what I was afraid of - I don't think we test the part of the wallet code that is used by the GUI coin selection very well.

**Comment by instagibbs** [comment]
did some additional refactoring to make flow easier to follow(early return if you captured enough value with presets), and added a basic unit test that should have failed with previous broken iteration

File: src/wallet/test/coinselector_tests.cpp [hunk 2]
```diff
@@ -236,6 +236,28 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
...
+    coin_control2.Select(COutPoint(vCoins.at(1).tx->GetHash(), vCoins.at(1).i));
+    coin_control2.Select(COutPoint(vCoins.at(2).tx->GetHash(), vCoins.at(2).i));
+    BOOST_CHECK(testWallet.SelectCoins(vCoins, 4 * CENT, setCoinsRet, nValueRet, coin_control2, coin_selection_params_bnb, bnb_used));
+    BOOST_CHECK(!bnb_used);
+    BOOST_CHECK(!coin_selection_params_bnb.use_bnb);
+    std::vector<COutPoint> coin_list;
+    coin_control2.ListSelected(coin_list);
+    BOOST_CHECK(setCoinsRet.size() == coin_list.size());
+    // "over-selected" since these were mandatory
+    BOOST_CHECK(nValueRet == (5+6+11) * CENT);
```
Comment (Empact) [inline src/wallet/test/coinselector_tests.cpp]: `BOOST_CHECK_EQUAL` here and above?

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2554,26 +2540,35 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
...
+        nValueRet = nValueFromPresetInputs;
+        setCoinsRet = setPresetCoins;
+        return (nValueRet >= nTargetValue);
+    }
+
     // remove preset inputs from vCoins
-    for (std::vector<COutput>::iterator it = vCoins.begin(); it != vCoins.end() && coin_control.HasSelected();)
-    {
-        if (setPresetCoins.count(CInputCoin(it->tx->tx, it->i)))
+    for (std::vector<COutput>::iterator it = vCoins.begin(); it != vCoins.end() && coin_control.HasSelected();) {
```
Comment (Empact) [inline src/wallet/wallet.cpp]: nit: this could be a const_iterator

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2554,26 +2540,35 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
...
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, 2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, std::min((size_t)4, nMaxChainLength/3)), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, nMaxChainLength/2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, nMaxChainLength), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && !fRejectLongChains && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max()), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used));
+    CAmount amount_to_select = nTargetValue - nValueFromPresetInputs;
```
Comment (Empact) [inline src/wallet/wallet.cpp]: nit: const

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: no, it's deleting entries?

Comment (Empact) [reply src/wallet/wallet.cpp]: The (non-const) vector is deleting entries, the COutputs themselves are unmodified.

Comment (instagibbs) [reply src/wallet/wallet.cpp]: ah, I assume const_iterator was const for everything, including the underlying vector. TIL

**Comment by instagibbs** [comment]
rebased, and did a tiny bit more refactoring for more readability.

**Comment by Empact** [comment]
You have some whitespace inconsistencies, clang-format would be good.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -2576,27 +2562,39 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
...
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, std::min((size_t)4, max_ancestors/3), std::min((size_t)4, max_descendants/3)), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, max_ancestors/2, max_descendants/2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) ||
-        (m_spend_zero_conf_change && !fRejectLongChains && SelectCoinsMinConf(nTargetValue - nValueFromPresetInputs, CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max()), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used));
+    const CAmount amount_to_select = nTargetValue - nValueFromPresetInputs;
```
Comment (Empact) [inline src/wallet/wallet.cpp]: This loses the early return if `nTargetValue <= nValueFromPresetInputs`

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -2576,27 +2562,39 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
...
+        SelectCoinsMinConf(amount_to_select, CoinEligibilityFilter(1, 1, 0), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used);
+
+    if (!res && m_spend_zero_conf_change) {
+             res = SelectCoinsMinConf(amount_to_select, CoinEligibilityFilter(0, 1, 2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used) ||
+            SelectCoinsMinConf(amount_to_select, CoinEligibilityFilter(0, 1, std::min((size_t)4, max_ancestors/3), std::min((size_t)4, max_descendants/3)), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used) ||
+            SelectCoinsMinConf(amount_to_select, CoinEligibilityFilter(0, 1, max_ancestors/2, max_descendants/2), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used) ||
+            SelectCoinsMinConf(amount_to_select, CoinEligibilityFilter(0, 1, max_ancestors-1, max_descendants-1), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used) ||
+            (!fRejectLongChains && SelectCoinsMinConf(amount_to_select, CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max()), vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used));
+    }
```
Comment (Empact) [inline src/wallet/wallet.cpp]: nit: I like not having redundant tests of `m_spend_zero_conf_change`, but I would keep it as a single assignment.

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: should be covered here: https://github.com/bitcoin/bitcoin/pull/13057/files#diff-b2bb174788c7409b671c46ccc86034bdR2567

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (instagibbs) [reply src/wallet/wallet.cpp]: done

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (Empact) [reply src/wallet/wallet.cpp]: Ah, thanks, in that case how about moving this assignment up to tie the two together?

Comment (instagibbs) [reply src/wallet/wallet.cpp]: done

**Comment by instagibbs** [comment]
rebased

**Comment by Empact** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/13057/commits/4f0111ecd6a9521da68b7597c73ffd7aa8bef98d but this would be easier to review if it focused around the fix https://github.com/Empact/bitcoin/commit/3bcd12d21f9198541a74422d186cf7cd2cefad49

**Comment by laanwj** [comment]
Any specific reason for closing this?

**Comment by instagibbs** [comment]
No particular reason. Just something I didn't feel like pursuing review on.

**Pull request closed without merging**